elastic/logstash
# [Spacetime] Reimplement config Setting classe in java

## Release notes

[rn:skip] 


## What does this PR do?
Reimplement the root Ruby Setting class in Java and use it from the Ruby one moving the original Ruby class to a shell wrapping the Java instance.
In particular create a new symmetric hierarchy (at the time just for Setting, Coercible and Boolean classes) to the Ruby one, moving also the feature for setting deprecation. In this way the new `org.logstash.settings.Boolean` is syntatically and semantically equivalent to the old Ruby Boolean class, which replaces. 


## Why is it important/What is the impact to the user?
This PR allows to move the Ruby Setting subclasses one at a time without breaking existing code.

## Checklist

## Author's Checklist
- [x] Fix the Ruby side of hierarchy to have green tests
 
## How to test this PR locally
The test consist in switching on and off a boolean setting and check the selection is respected. Consider the  `api.enabled` and its deprecated version named `http.enabled`, tinker with it in `config/logstash.yml` and as command line switch `--api.enabled true|false`. Check with `curl`.

1. set to false `api.enabled` in `logstash.yml`
2. run Logstash. with `bin/logstash -e "input{stdin{}} output{stdout{}}"`
3. verify with
```sh
curl -v "http://localhost:9600"
```
4. Use the command line switch 
```sh
bin/logstash -e "input{stdin{}} output{stdout{}}" --api.enabled false 
```
## Related issues


- Closes #12531
- Closes #12566

## Use cases



## Screenshots



## Logs
